provider/pkg/provider/yaml/v2/yaml.go

```diff
@@ -0,0 +1,259 @@
+			continue
+		}
+		resources = append(resources, resource)
+	}
+
+	result := make([]map[string]any, 0, len(resources))
+	for _, resource := range resources {
+		result = append(result, resource.Object)
+	}
+	return result, nil
+}
+
+func ParseYamlObjects(
+	ctx *pulumi.Context,
+	objs []map[string]any,
+	resourcePrefix string,
+	skipAwait bool,
+	opts ...pulumi.ResourceOption,
+) (pulumi.ArrayOutput, error) {
+	var intermediates []pulumi.CustomResource
+	for _, obj := range objs {
+		rs, err := parseYamlObject(ctx, obj, resourcePrefix, skipAwait, opts...)
+		if err != nil {
+			return pulumi.ArrayOutput{}, err
+		}
+		intermediates = append(intermediates, rs...)
+	}
+
+	resources := pulumi.Array{}
+	for _, r := range intermediates {
+		resources = append(resources, pulumi.NewResourceOutput(r))
+	}
+	return resources.ToArrayOutputWithContext(ctx.Context()), nil
+}
+
+func parseYamlObject(
+	ctx *pulumi.Context,
+	obj map[string]any,
+	resourcePrefix string,
+	skipAwait bool,
+	opts ...pulumi.ResourceOption,
+) ([]pulumi.CustomResource, error) {
+
+	getProp := func(key string) (string, bool) {
+		v, ok := obj[key]
+		if !ok {
+			return "", false
+		}
+		return fmt.Sprintf("%s", v), true
+	}
+
+	// Ensure there is a kind and API version.
+	kind, hasKind := getProp("kind")
+	apiVersion, hasAPIVersion := getProp("apiVersion")
+	if !hasKind || !hasAPIVersion {
+		return nil, errors.Errorf("Kubernetes resources require a kind and apiVersion: %+v", obj)
+	}
+	fullKind := fmt.Sprintf("%s/%s", apiVersion, kind)
+
+	// Recursively traverse built-in Kubernetes list types into a single set of "naked" resource
+	// definitions that we can register with the Pulumi engine.
+	//
+	// Kubernetes does not instantiate list types like `v1.List`. When the API server receives
+	// a list, it will recursively traverse it and perform the necessary operations on
+	// each "instantiable" resource it finds. For example, `kubectl apply` on a
+	// `v1.ConfigMapList` will cause the API server to traverse the list, and `apply` each
+	// `v1.ConfigMap` it finds.
+	//
+	// Since Kubernetes does not instantiate list types directly, Pulumi also traverses lists
+	// for resource definitions that can be managed by Kubernetes, and registers those with the
+	// engine instead.
+	if yamlv2.IsListKind(apiVersion, kind) {
+		var resources []pulumi.CustomResource
+		if rawItems, hasItems := obj["items"]; hasItems {
+			if items, ok := rawItems.([]any); ok {
+				for _, item := range items {
+					if obj, ok := item.(map[string]any); ok {
+						rs, err := parseYamlObject(ctx, obj, resourcePrefix, skipAwait, opts...)
+						if err != nil {
+							return nil, err
+						}
+						resources = append(resources, rs...)
+					}
+				}
+			}
+		}
+		return resources, nil
+	}
+
+	// If we got here, it's not a recursively traversed type, so process it directly.
+	// First, validate that it has the requisite metadata and name properties.
+	meta, hasMeta := obj["metadata"]
```

Comment:
WDYT about using the helper functions from the unstructured library to work with these objects?

Reply:
An excellent idea, thanks.
https://pkg.go.dev/k8s.io/apimachinery@v0.29.0/pkg/apis/meta/v1/unstructured#pkg-index